python/ray/dag/compiled_dag_node.py

```diff
@@ -373,13 +385,53 @@ def _get_or_compile(
             self.dag_output_channels = self.dag_output_channels[0]
 
         # Driver should ray.put on input, ray.get/release on output
-        return (self.dag_input_channel, self.dag_output_channels)
+        self._monitor = self._monitor_failures()
+        return (self.dag_input_channel, self.dag_output_channels, self._monitor)
+
+    def _monitor_failures(self):
+        outer = self
+
+        class Monitor(threading.Thread):
+            def __init__(self):
+                super().__init__(daemon=True)
+                self.in_teardown = False
+
+            def teardown(self):
+                if self.in_teardown:
+                    return
+                logger.info("Tearing down compiled DAG")
+                self.in_teardown = True
+                for actor in outer.actor_refs:
```

Comment:
Wouldn't there be thread safety issue here?  (if the dag accesses actor_refs while monitor thread is calling teardown)

maybe we should lock this field?

Reply:
What do you mean? I don't think this field is mutated after creation.

Reply:
i think in_teardown is okay (bool is atomically updated in python iirc), but actor_ref and worker_task_ref are now accessed by 2 threads. 

It is bad to assume implicit behavior that these two are not modified imo (although it may be true at this point). Someone can add some code later and screw things up easily. Maybe at least can you comment on outer.actor ref & outer.worker_task ref that they must be immutable?